src/main/java/ch/njol/skript/lang/function/DefaultFunction.java

```diff
@@ -0,0 +1,341 @@
+package ch.njol.skript.lang.function;
+
+import ch.njol.skript.classes.ClassInfo;
+import ch.njol.skript.registrations.Classes;
+import com.google.common.base.Preconditions;
+import org.jetbrains.annotations.Contract;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.lang.reflect.Array;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * A function that has been implemented in Java, instead of in Skript.
+ * <p>
+ * An example implementation is stated below.
+ * <pre><code>
+ * Functions.register(DefaultFunction.builder("floor", Long.class)
+ * 	.description("Rounds a number down.")
+ * 	.examples("floor(2.34) = 2")
+ * 	.since("3.0")
+ * 	.parameter("n", Number.class)
+ * 	.build(args -> {
+ * 		Object value = args.get("n");
+ *
+ * 		if (value instanceof Long l)
+ * 			return l;
+ *
+ * 		return Math2.floor(((Number) value).doubleValue());
+ *    }));
+ * </code></pre>
+ * </p>
+ *
+ * @param <T> The return type.
+ * @see #builder(String, Class)
+ */
+public final class DefaultFunction<T> extends ch.njol.skript.lang.function.Function<T> {
+
+	/**
+	 * Creates a new builder for a function.
+	 *
+	 * @param name       The name of the function.
+	 * @param returnType The type of the function.
+	 * @param <T>        The return type.
+	 * @return The builder for a function.
+	 */
+	public static <T> Builder<T> builder(@NotNull String name, @NotNull Class<T> returnType) {
+		return new Builder<>(name, returnType);
+	}
+
+	private final Parameter<?>[] parameters;
+	private final Function<FunctionArguments, T> execute;
+
+	private final String[] description;
+	private final String[] since;
+	private final String[] examples;
+	private final String[] keywords;
+
+	private DefaultFunction(
+		String name, Parameter<?>[] parameters,
+		ClassInfo<T> returnType, boolean single,
+		@Nullable ch.njol.skript.util.Contract contract, Function<FunctionArguments, T> execute,
+		String[] description, String[] since, String[] examples, String[] keywords
+	) {
+		super(new Signature<>("none", name, parameters, false,
```

Comment:
just to note that this will need updated for overloaded functions (in which ever PR is merged last)